Drop use of the Intl API to improve browser support

ocombe opened the pull request:
## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[x] Feature
```

## What is the current behavior?
We need the intl API for i18n pipes

Issue Number: #10809, #9524, #7008, #9324, #7590, #6724, #3429, #17576, #17478, #17319, #17200, #16838, #16624, #16625, #16591, #14131, #12632, #11376, #11187


## What is the new behavior?
We don't need the intl API anymore, we extract and use data from CLDR instead.

### Date pipe:
- new formats (breaking change), see [comparative document](https://docs.google.com/spreadsheets/d/12iygt-_cakNP1VO7MV9g4lq9NsxVWG4tSfc98HpHb0k/edit?usp=sharing)
- new optional parameter: `timezone` that lets you specify the timezone for a date (it will shift the date values based on the locale timezone so that you see it as it was from this other timezone).
- new optional parameter: `locale` that lets you use a specific locale for this pipe (instead of using the default `LOCALE_ID`)
 
### Plural pipe:
- new optional parameter: `locale` that lets you use a specific locale for this pipe (instead of using the default `LOCALE_ID`)
 
### Decimal pipe:
- new optional parameter: `locale` that lets you use a specific locale for this pipe (instead of using the default `LOCALE_ID`)

### Percent pipe:
- new optional parameter: `locale` that lets you use a specific locale for this pipe (instead of using the default `LOCALE_ID`)
 
### Currency pipe:
- the parameter `symbolDisplay` is no longer a boolean (it still works with a boolean, but it prints a warning to let you know that it changed), it’s now taking the values `code`, `symbol`, or `symbol-narrow` (the default is still code) which gives you access to more options for to show your currencies (e.g. the canadian dollar with the code `CAD` has the symbol `CA$` and the symbol-narrow `$`).
- new optional parameter: `locale` that lets you use a specific locale for this pipe (instead of using the default `LOCALE_ID`)

## Does this PR introduce a breaking change?
```
[x] Yes
```

All i18n pipes have been updated to not require the intl API anymore. Some results might vary, see the document to update from v4 to v5 for more information.

## TODO
- write a document to update from v4 to v5

ocombe committed:
build(common): extract i18n locale data from cldr

ocombe committed:
build(common): add generated i18n locale data files

mary-poppins commented:
You can preview 38d3eed at https://pr18284-38d3eed.ngbuilds.io/.

vicb commented on tools/gulp-tasks/cldr/extract.js:
```diff
@@ -0,0 +1,398 @@
+
+  const dateTimeTranslations = {dayPeriodsFormat, daysFormat, monthsFormat, eras};
+
+  if(dayPeriodsFormat !== dayPeriodsStandalone) {
```
add comment on dedup

vicb commented on packages/common/src/i18n/localization.ts:
```diff
@@ -0,0 +1,127 @@
+ * ie: "en-US" will return "en", and "fr_ca" will return "fr-CA"
+ */
+function getNormalizedLocale(locale: string): string {
+  const normalizedLocale = locale.toLowerCase().replace(new RegExp('_', 'g'), '-');
```
nit: `new RegExp('_', 'g')` -> `/-/g`

vicb commented on packages/common/src/i18n/localization.ts:
```diff
@@ -0,0 +1,127 @@
+ * @experimental i18n support is experimental.
+ */
+export function findNgLocale(locale: string, localeData: NgLocale[] | null): NgLocale {
+  const currentLocale = getNormalizedLocale(locale);
```
currentLocale -> normalizedLocale

vicb commented on packages/platform-browser/test/browser_util_spec.ts:
```diff
@@ -234,7 +234,6 @@ export function main() {
         expect(bd.isIOS7).toBe(browser['isIOS7']);
         expect(bd.isSlow).toBe(browser['isSlow']);
         expect(bd.isChromeDesktop).toBe(browser['isChromeDesktop']);
-        expect(bd.isOldChrome).toBe(browser['isOldChrome']);
```
why this change ?

ocombe replied:
`isOldChrome` is only used by intl api, when I first did the PR, I removed the old pipes and the tests for them. But then I moved them as deprecated pipes instead, and forgot to revert this change. Doing it now.

mary-poppins commented:
You can preview 1d5eed7 at https://pr18284-1d5eed7.ngbuilds.io/.

mary-poppins commented:
You can preview 1053613 at https://pr18284-1053613.ngbuilds.io/.

mary-poppins commented:
You can preview d513368 at https://pr18284-d513368.ngbuilds.io/.

mary-poppins commented:
You can preview c33247f at https://pr18284-c33247f.ngbuilds.io/.

mary-poppins commented:
You can preview 052a61a at https://pr18284-052a61a.ngbuilds.io/.

vicb commented on packages/common/i18n/full/locale_af-NA.ts:
```diff
@@ -0,0 +1,68 @@
+
+import {NgLocaleAfNA} from '../locale_af-NA';
+
+const dpF = {
```
change `dpF

vicb commented on tools/gulp-tasks/cldr/extract.js:
```diff
@@ -0,0 +1,450 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+// This is generated code DO NOT MODIFY
```
THIS CODE IS GENERATED - DO NOT MODIFY

vicb commented on packages/common/i18n/full/locale_af-NA.ts:
```diff
@@ -0,0 +1,68 @@
+const format = NgLocaleAfNA.dateTime.translations.dayPeriods.format;
+const std = NgLocaleAfNA.dateTime.translations.dayPeriods.standalone;
+
+format.narrow = {...format.narrow, ...dpF.narrow};
```
Thoughts:
- I think we can remove all of the `{..., ...}`,
- I think we can also remove the need to import the std objects and merge the full object

Because we are going to use an API to access the data (which goal is to abstract the storage format).

Also the `registerLocaleData(...)` could help here.

vicb commented on tools/gulp-tasks/cldr/extract.js:
```diff
@@ -0,0 +1,450 @@
+    settings: ${getDateTimeSettings(localeData)}
+  },
+  number: {
+    settings: ${getNumberSettings(localeData)}
```
do we need a nested key here and below ?

ocombe replied:
not yet, just thinking about future things that we might add

vicb commented on tools/gulp-tasks/cldr/extract.js:
```diff
@@ -0,0 +1,450 @@
+  const dpF = stringify(dayPeriods.format);
+  let dpStd = stringify(dayPeriods['stand-alone']);
+
+  if (dpF === dpStd) {
```
add a comment explaining what you do pls

vicb commented on tools/gulp-tasks/cldr/extract.js:
```diff
@@ -0,0 +1,450 @@
+
+const dpF = ${dpF};${dpStd ? `\nconst dpStd = ${dpStd};` : ''}
+
+const format = NgLocale${toPascalCase(locale)}.dateTime.translations.dayPeriods.format;
```
There is probably no need to duplicate this logic as many time as we have locales

vicb commented on tools/gulp-tasks/cldr/extract.js:
```diff
@@ -0,0 +1,450 @@
+  const dayPeriodsStandalone = stringify(dayPeriods['stand-alone']);
+
+  const daysFormat = stringify({
+    narrow: objectValues(dayNames.format.narrow),
```
Have you tried to generate arrays instead of obj ?
What is the diff in size ?

vicb commented on packages/common/i18n/full/locale_ar-QA.ts:
```diff
@@ -0,0 +1,94 @@
+/**
```
Idea: I donc speak `ar` but the files look really similar for different regions (the `QA` part). Is there any check in place to make sure we do not duplicate the same data multiple times ?

ocombe replied:
Not between different locales no, but it doesn't really matter since people will only import one of them, not all

vicb replied:
It will prevent devs to have to buy a new SSD only to npm install Angular. 
The check is easy enough. Please git it a try to see how much we can save.

vicb commented on packages/common/i18n/locale_fr-BJ.ts:
```diff
@@ -0,0 +1,98 @@
+// This is generated code DO NOT MODIFY
+// See angular/tools/gulp-tasks/cldr/extract.js
+
+import {LOCALE_DATA, NgLocale, Plural} from '@angular/common';
```
meme remarque encore ici: probablement beaucoup de chose en commun avec `fr-*` (sauf currency et peut etre numbers). Would be NICE to check what we can factor.

vicb commented on packages/common/src/i18n/currencies.ts:
```diff
@@ -0,0 +1,312 @@
+
+/** @experimental */
+export const CURRENCIES: {[code: string]: {[key: string]: string}} = {
+  'ADP': {symbol: 'ADP'},
```
An array would make sense here instead of an obj.

ocombe replied:
👍 Just saved 8.5ko minified by optimizing this file!

vicb commented on packages/common/i18n/tsconfig-build.json:
```diff
@@ -0,0 +1,21 @@
+      "@angular/common": ["../../../dist/packages/common"],
+      "@angular/core": ["../../../dist/packages/core"]
+    },
+//    "rootDir": ".",
```
DEBUG ?

vicb commented on packages/common/src/i18n/ng_locale.ts:
```diff
@@ -0,0 +1,154 @@
+// see http://cldr.unicode.org/translation/date-time#TOC-Stand-Alone-vs.-Format-Styles
+// Field  format	standAlone
+// Month	M     	L
+// Day    E	      c
```
??

vicb commented on tools/gulp-tasks/cldr/extract.js:
```diff
@@ -0,0 +1,457 @@
+    const rules = [];
+
+    keys.forEach(key => {
+      rules.push(dayPeriodRules[key]);
```
`const rules = keys.map();`

vicb commented on packages/common/src/i18n/locale_en.ts:
```diff
@@ -0,0 +1,71 @@
+LOCALE_DATA['en'] = [
+    'en',
+    [
+      [ [ [ [ 'a', 'p' ], [ 'AM', 'PM' ], [ 'AM', 'PM' ] ],
```
Do we need that much nested arrays ?

ocombe replied:
I'll try to remove some of them

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -0,0 +1,358 @@
+ * @internal
+ */
+export function formatNumber(
+    pipe: Type<any>, value: number | string, style: NumberFormatStyle, format: string,
```
pipe ???

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -0,0 +1,358 @@
+  // Convert strings to numbers
+  number = typeof value === 'string' && isNumeric(value) ? +value : value as number;
+  if (typeof number !== 'number') {
+    throw invalidPipeArgumentError(pipe, number);
```
pipe ???

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -0,0 +1,358 @@
+    if (digitsInfo) {
+      const parts = digitsInfo.match(NUMBER_FORMAT_REGEXP);
+      if (parts === null) {
+        throw new Error(`${digitsInfo} is not a valid digit info for number pipes`);
```
pipes

vicb commented on packages/common/src/i18n/locale_api.ts:
```diff
@@ -0,0 +1,152 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+const FORMS = ['format', 'standalone'];
```
any reason not to use enums ?

ocombe replied:
no, good idea

vicb commented on packages/common/src/i18n/locale_api.ts:
```diff
@@ -0,0 +1,152 @@
+ *
+ * @experimental i18n support is experimental.
+ */
+export class I18nApi {
```
Thoughts:
Does it really makes sense to make it a class ?

ocombe replied:
Well it uses the constructor to get the data. This way you don't need to know how to get the data (from `LOCALE_DATA`), otherwise it means that you need to pass the locale data to each function, which means that you need to get them in each pipe / service that will use this api.

vicb commented on packages/common/src/i18n/locale_api.ts:
```diff
@@ -0,0 +1,152 @@
+  localeId(): string { return this.data[0]; }
+
+  amPm(form: Form, width: TranslationWidth): [string, string] {
+    const amPm = this.data[1][0][0];
```
`[1][0][0]`-> constants

vicb commented on packages/common/src/i18n/locale_api.ts:
```diff
@@ -0,0 +1,152 @@
+    if (form === 'standalone') {
+      data = amPm[1] || data;
+    }
+    return data[TRANSLATION_WIDTHS.indexOf(width)];
```
-> with numeric enums you should be able to remove `indexOf`

vicb commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -117,30 +76,35 @@ export class DecimalPipe implements PipeTransform {
 @Pipe({name: 'percent'})
 export class PercentPipe implements PipeTransform {
-  constructor(@Inject(LOCALE_ID) private _locale: string) {}
+  constructor(@Inject(LOCALE_ID) private locale: string) {}
```
ping

vicb commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -88,26 +45,28 @@ function formatNumber(
 @Pipe({name: 'number'})
 export class DecimalPipe implements PipeTransform {
-  constructor(@Inject(LOCALE_ID) private _locale: string) {}
+  constructor(@Inject(LOCALE_ID) private locale: string) {}
```
ping

mary-poppins commented:
You can preview 5940f28 at https://pr18284-5940f28.ngbuilds.io/.

vicb commented on packages/common/src/common_module.ts:
```diff
@@ -29,3 +30,12 @@ import {COMMON_PIPES} from './pipes/index';
+/**
+ * A module to contain deprecated i18n pipes.
+ *
+ * @deprecated
```
`@deprecated` -> from when

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+ *
+ * @internal
+ */
+export function formatDatePattern(
```
api as a class again - to discuss

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -0,0 +1,369 @@
+ * @internal
+ */
+export function formatNumber(
+    value: number | string, style: NumberFormatStyle, format: string, api: LocaleDataApi,
```
- api again to discuss,
- argument order is not consistent with the date counterpart

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -0,0 +1,369 @@
+    lgSize: 0
+  };
+
+  const patternParts = format.split(PATTERN_SEP), positive = patternParts[0],
```
we don't use coma in the CS. 
Use multiple const

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,349 @@
+   * First day of the week for this locale, based on english days (Sunday = 0, Monday = 1, ...).
+   * For example in french the value would be 1 because the first day of the week is Monday.
+   */
+  firstDayOfWeek(): number { return this.data[8]; }
```
We should return an enum `enum Days { Sunday = 0, ...}`

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,349 @@
+   *
+   *  For example, `M/d/yy`, corresponding to a date like "9/14/99".
+   */
+  dateFormat(width: FormatWidth): string { return this.data[10][width]; }
```
`this.data[10]` -> pls use constants

ocombe replied:
Constant for "10" or for "this.data[10]" ?
Something like `this.data[DATE_FORMAT]` where `DATE_FORMAT = 10` ? I could even use an enum here actually

vicb replied:
10

mary-poppins commented:
You can preview f2d1c45 at https://pr18284-f2d1c45.ngbuilds.io/.

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,349 @@
+}
+
+function firstDefined(data: any[], index: number): any {
+  let res, i = index;
```
- do not use `,`
- a `for` loop might be clearer
- `!== undefined` would be be better (and does what the function name implies)

ocombe replied:
Actually `!== undefined` was necessary, otherwise we could have an error when res is "" or 0 for example. Good catch.

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+export function formatDatePattern(
+    api: LocaleDataApi, date: Date, pattern: string, timezone?: string): string {
+  const shortTime = api.timeFormat(FormatWidth.Short);
+  const specialPatterns: {[key: string]: string} = {
```
- name = "namedPatterns" ?
- it should probably not be a local variable (re-created each time the fn is called)
- it should probably not be an object (we don't want to compute all the value each time we init the obj)
- same comment for below

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+
+/** @internal */
+export function toInt(str: string): number {
+  return parseInt(str, 10);
```
do we really need that ? and to export that ?

ocombe replied:
It's only exported because it's used in multiple internal files, but it stays internal. Using this function probably gains a bit of space in the final bundle? It was like this in angularjs, I don't mind replacing by parseInt everywhere it is used, if there is a point to it?

vicb replied:
the points are that:
- there is no reason why `format_date` should export a `toInt`, who would look for it there ?
- closure takes care of optimizing the size (much better than any other tricks)

ocombe replied:
Most people don't use closure, but I guess it's the same with uglify & co

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+      value += offset;
+    }
+    if (value === 0 && offset === -12) {
+      value = 12;
```
what is that ?

ocombe replied:
I explained in the previous PR, but it disappeared.
It's for AM/PM --> 00h00 = 12 am (at least that's how I understand it).

vicb replied:
> this comes from angularjs
This is not a good reason to introduce a bug.

This is probably missing `name === DateType.Hours &&` (& please also fix angular.js)

ocombe replied:
Well the offset argument is only used for hours in AM/PM , so there is no need to fix angularjs (it's not like they are going to change the date pipe), but I'll add a check here just in case

vicb commented on packages/common/src/pipes/date_pipe.ts:
```diff
@@ -170,6 +204,12 @@ function isoStringToDate(match: RegExpMatchArray): Date {
-function toInt(str: string): number {
-  return parseInt(str, 10);
+/** @internal */
+export function isBlank(obj: any): boolean {
```
pipes should not export utils function

ocombe replied:
I only export it because it's also used in the deprecated pipe and I wanted to remove the duplicated code, do you prefer to have the same function in both pipes?

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+function dateGetter(
+    name: DateType, size: number, offset: number = 0, trim = false,
+    negWrap = false): DateFormatter {
+  return function(date: Date, api: LocaleDataApi): string {
```
breakdown to improve readability:

```ts
return function((...) {
  part = getDatePart(...);
  // offset black magic
  return padNumber(...);
}

----

function getDatePart {
  switch {
    case a : return A;
    case b: return B;
 }

}

```

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+    name: TranslationType, width: TranslationWidth, form: FormStyle = FormStyle.Format,
+    extended = false): DateFormatter {
+  return function(date: Date, api: LocaleDataApi): string {
+    switch (name) {
```
same comment -> same refactoring as above

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+    switch (width) {
+      case ZoneWidth.Short:
+        value = ((zone >= 0) ? '+' : '') +
+            padNumber(Math[zone > 0 ? 'floor' : 'ceil'](zone / 60), 2, minusSign) +
```
please use `zone > 0 ? Math.floor(...) : ...` better for minification & static analysis.

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+        if (offset === 0) {
+          value = 'Z';
+        } else {
+          // TODO(ocombe): support optional seconds field, if there really is a use case
```
`if there really is a use case` -> remove the TODO for now

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+}
+
+function getFirstThursdayOfYear(year: number) {
+  // 0 = index of January
```
use a constant instead of number + comment ?

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+// http://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table
+// See also explanations: http://cldr.unicode.org/translation/date-time
+// TODO(ocombe): support all missing cldr formats: Y, U, Q, D, F, e, c, j, J, C, A, v, V, X, x
+const DATE_FORMATS: {[format: string]: DateFormatter} = {
```
What about computing this lazily ?

ocombe replied:
Do you mean only creating "DATE_FORMATS" the first time we need it? If you want, I'm not sure about the gain, we probably want the date pipe to be as fast as possible when it runs since it's synchronous, but we also compute so many things at bootstrap that removing this from the critical path might be better, especially if we never use the date pipe.

vicb replied:
creating `DATE_FORMATS.[format]` - additional benefit is that most apps probably only need a small subset

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,349 @@
+   * Periods of the day (e.g. AM/PM).
+   */
+  amPm(formStyle: FormStyle, width: TranslationWidth): [string, string] {
+    let amPm = firstDefined([this.data[1], this.data[2]], formStyle);
```
const

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,349 @@
+   * Days of the week for the Gregorian calendar (e.g. `Sunday`).
+   */
+  days(formStyle: FormStyle, width: TranslationWidth): string[] {
+    let months = firstDefined([this.data[3], this.data[4]], formStyle);
```
const

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,349 @@
+   * (Sunday = 0, Monday = 1, ...).
+   * For example in english the value would be [6,0] for Saturday to Sunday.
+   */
+  weekEndRange(): [number, number] { return this.data[9]; }
```
same thing -> enum

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,349 @@
+  weekEndRange(): [number, number] { return this.data[9]; }
+
+  /**
+   * There are four basic date formats.
```
It's good to have API docs but reading this I have no clue what the method returns ?

replace `full` with `FormatWidth.full`, ... ?

ocombe replied:
good point, I'll add example of the returned data

vicb replied:
>  I'll add example of the returned data

I don't think that much is missing.

But API docs *should* start with a sentence describing what the method/function/class does. This is missing here.

vicb commented on packages/common/src/pipes/date_pipe.ts:
```diff
@@ -65,8 +116,8 @@ const ISO8601_DATE_REGEX =
  *   which would be an expensive operation).
- * - this pipe uses the Internationalization API. Therefore it is only reliable in Chrome and Opera
- *   browsers.
+ * - this pipe only contains en-US locale data by default. You have to import data for other locales
```
`this pipe only contains en-US` -> rephrase

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -0,0 +1,369 @@
+/** @internal */
+export type FormatNumberRes = {
+  str: string | null,
+  error: string | null
```
should probably be `error?:string` - same for `str` ?

ocombe replied:
str should as least be `string | null` because the number pipes can return `string | null`, not undefined

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -0,0 +1,369 @@
+    value: number | string, style: NumberFormatStyle, format: string, api: LocaleDataApi,
+    digitsInfo?: string | null, currency: string | null = null): FormatNumberRes {
+  const res: FormatNumberRes = {str: null, error: null};
+  if (value == null) return res;
```
should it be an error ? or fix the param types ?

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -0,0 +1,369 @@
+  let number: number;
+
+  // Convert strings to numbers
+  number = typeof value === 'string' && isNumeric(value) ? +value : value as number;
```
`number = value as number;` is weird when value is not a number.

refactor 
```
if (value is numeric string) {
  value = +value;
} 
if (typeof value !== 'number') {
  return ...;
}
// TS should infer that value is a number from here

```


```

vicb commented:
There is a circular dep (check Travis):

```
Found circular dependencies!
[ [ '@angular/common/src/i18n/localization',
    '@angular/common/src/i18n/locale_data_api' ] ]
```

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,502 @@
+  EEEE: dateStrGetter(TranslationType.Days, TranslationWidth.Wide),
+  // Day in Week narrow (S-S)
+  EEEEE: dateStrGetter(TranslationType.Days, TranslationWidth.Narrow),
+  // Day in Week narrow (Su-Sa)
```
typo (not narrow).

Also with enum a single comment on the group might be enough ? (ie `// Day in week`)

mary-poppins commented:
You can preview 101e17d at https://pr18284-101e17d.ngbuilds.io/.

mary-poppins commented:
You can preview a4d4596 at https://pr18284-a4d4596.ngbuilds.io/.

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -112,72 +171,54 @@ function padNumber(
+function getDatePart(name: DateType, date: Date, size: number, offset: number): number {
+  let value = 0;
+  switch (name) {
+    case DateType.FullYear:
```
check the former comment:

```
case DateType.Month:
     return date.getMonth();
```

-> `getDartPart()` is expected to return a date part. Handle offset at calling site.

Throw on case default.

vicb replied:
again:
```
case DateType.Month:
     return date.getMonth();
```

no need for value & break

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -277,19 +318,17 @@ function timeZoneGetter(width: ZoneWidth): DateFormatter {
-  // 11 = index of *next* Thursday (+1 account for 1st = 12)
-  return new Date(year, 0, ((dayOfWeekOnFirst <= 4) ? 5 : 12) - dayOfWeekOnFirst);
+  const firstDayOfYear = (new Date(year, JANUARY, 1)).getDay();
+  return new Date(year, 0, ((firstDayOfYear <= THURSDAY) ? 5 : 12) - firstDayOfYear);
```
5 ? 12 ?
-> THURSDAY + 1 / THURSDAY + 1 + 7

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -312,173 +351,246 @@ function weekGetter(size: number, monthBased = false): DateFormatter {
+    // Should be location, but fallback to format O instead because we don't have the data yet
+    case 'zzzz':
+      formatter = timeZoneGetter(ZoneWidth.Long);
+      break;
```
default: return null;

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -312,173 +351,246 @@ function weekGetter(size: number, monthBased = false): DateFormatter {
+      formatter = timeZoneGetter(ZoneWidth.Long);
+      break;
+  }
+  if (formatter) {
```
remove the if

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -20,7 +21,7 @@ const PERCENT_CHAR = '%';
 /** @internal */
 export type FormatNumberRes = {
   str: string | null,
```
when/why null ?

ocombe replied:
because the pipes require string or null, not undefined

vicb replied:
Please add a comment on the type, ie when is this `null` ? (Can we assume this is not null if there is no error).

_There is no reason why one consumer (the pipe) should dictate the data format_

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -111,30 +113,44 @@ export enum NumberSymbol {
+ * @experimental
+ */
+export enum WeekDay {
+  Sunday,
```
we want to be explicit with day index, don't we (check my former comment)

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -111,30 +113,44 @@ export enum NumberSymbol {
+}
+
+enum LocaleData {
+  LocaleId,
```
Should be the same here ?

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -255,95 +280,197 @@ export class LocaleDataApi {
+    res = data[i];
   }
-  return res;
+  return res !;
```
why `!` here ?

ocombe replied:
It cannot return undefined based on how I extract the data, which means that I don't want to change the api return type to include undefined, but given how the for loop works typescript thinks that it could be undefined

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -255,95 +280,197 @@ export class LocaleDataApi {
+function firstDefined<T>(data: T[], index: number): T {
+  let res;
+  for (let i = index; i > -1 && typeof res === 'undefined'; i--) {
+    res = data[i];
```
would be more clear to return early if not undefined inside the for, no ?

vicb commented on packages/common/src/util.ts:
```diff
@@ -0,0 +1,43 @@
+ *
+ * @internal
+ */
+export function isBlank(value: any): boolean {
```
We have spent time to remove all the facades we had - are those really helpful ?

mary-poppins commented:
You can preview 921c52b at https://pr18284-921c52b.ngbuilds.io/.

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -399,11 +399,12 @@ export class LocaleDataApi {
+    }
   }
-  return res !;
+  return data[index];
```
really ?

ocombe replied:
got a better idea?

ocombe replied:
we could also throw an error instead, since this use case is not supposed to happen

vicb replied:
yep this should throw

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -36,7 +36,7 @@ export function formatNumber(
 
   // Convert strings to numbers
-  if (typeof value === 'string' && isNumeric(value)) {
+  if (typeof value === 'string' && !isNaN(+value - parseFloat(value))) {
```
is the `+` needed / valid here ?

ocombe replied:
It's either that or `<any>` otherwise typescript throws (the function that you asked me to remove did cast it to any). I've done a few tests and it seems to be ok to use "+" ¯\_(ツ)_/¯

vicb commented on packages/common/src/i18n/localization.ts:
```diff
@@ -0,0 +1,131 @@
+
+/**
+ * Returns the closest matching locale that exists or throw
+ * ie: "en-US" will return "en", and "fr_ca" will return "fr-CA"
```
Not sure to understand the "en-US"->"en" transform ? Is it because "en-US" is not supposed to exist ?

(On a side note is it expected to have a "en-US-POSIX" locale ?)

mary-poppins commented:
You can preview e559e6f at https://pr18284-e559e6f.ngbuilds.io/.

mary-poppins commented:
You can preview 151f6ce at https://pr18284-151f6ce.ngbuilds.io/.

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,617 @@
+      value = date.getMonth();
+      break;
+    case DateType.Date:
+      value = date.getDate();
```
simplify to return value and remove the break

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,617 @@
+    const zone = -1 * offset;
+    const minusSign = api.numberSymbol(NumberSymbol.MinusSign);
+    let value = '';
+    switch (width) {
```
- compute `Math.floor(zone/60)` once before the switch
- early return instead of assigning to value

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,617 @@
+    } else {
+      const firstThurs = getFirstThursdayOfYear(date.getFullYear());
+      const thisThurs = getThursdayThisWeek(date);
+      const diff = +thisThurs - +firstThurs;
```
use `getTime` instead of `+`

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,477 @@
+  Saturday
+}
+
+enum LocaleData {
```
what is this ? (comment)

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,477 @@
+  constructor(public locale: string) { this.data = findNgLocale(locale); }
+
+  /**
+   * The locale id for the chosen locale, for example `en-US`.
```
Does it exist (see my other related question)

ocombe replied:
What do you mean by "does it exist" ?

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,477 @@
+  localeId(): string { return this.data[LocaleData.LocaleId]; }
+
+  /**
+   * Periods of the day (e.g. AM/PM).
```
> (e.g. AM/PM)

this method returns an array of string, how does this related ?? (same below)

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,477 @@
+  }
+
+  /**
+   * Days of the week for the Gregorian calendar (e.g. `Sunday`).
```
index 0 is ?

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,477 @@
+   * Here are the special characters used in number patterns:
+   *
+   * | Symbol | Meaning |
+   * |--------|---------|
```
pls use fixed width cols

ocombe replied:
I can't because it gets over the 100 characters limit and `gulp format` moves some of the text to the next line, which breaks the table in the docs... Is there a way to force gulp format to ignore a specific comment?

vicb replied:
`// clang-format [off/on]`

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,477 @@
+    return rules.map((rule: string | [string, string]) => {
+      if (typeof rule === 'string') {
+        return extractTime(rule);
+      } else {
```
no else

ocombe replied:
By the way do you think that it's ok if `dayPeriodRules` returns `Time|[Time, Time]` or should it return `Time[]` instead? (I could convert the single value to an array with only one value, either in the api or in the extract script)

mary-poppins commented:
You can preview 04c2c63 at https://pr18284-04c2c63.ngbuilds.io/.

mary-poppins commented:
You can preview 9cd9be3 at https://pr18284-9cd9be3.ngbuilds.io/.

mary-poppins commented:
You can preview 578d98a at https://pr18284-578d98a.ngbuilds.io/.

mary-poppins commented:
You can preview 1f70149 at https://pr18284-1f70149.ngbuilds.io/.

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,601 @@
+  return function(date: Date, api: LocaleDataApi, offset: number) {
+    const zone = -1 * offset;
+    const minusSign = api.numberSymbol(NumberSymbol.MinusSign);
+    const minutes = Math.floor(zone / 60);
```
hour ?

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,601 @@
+    const minutes = Math.floor(zone / 60);
+    switch (width) {
+      case ZoneWidth.Short:
+        return ((zone >= 0) ? '+' : '') + padNumber(zone > 0 ? minutes : minutes, 2, minusSign) +
```
`zone > 0 ? minutes : minutes`
test ?

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -0,0 +1,601 @@
+ * extended = +04:30)
+ */
+function timeZoneGetter(width: ZoneWidth): DateFormatter {
+  return function(date: Date, api: LocaleDataApi, offset: number) {
```
offset -> add unit

ocombe replied:
what do you mean?

vicb replied:
offset seconds ie

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,478 @@
+ *
+ * | Name | English Example | Meaning |
+ * |------|-----------------|---------|
+ * | decimal | 2,345`.`67 | decimal separator |
```
wrong

ocombe replied:
what is wrong?

vicb commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -74,15 +73,15 @@ export function formatDate(
     const dateFormatter = getDateFormatter(value);
     text += dateFormatter ?
-        dateFormatter(date, api, dateTimezoneOffset) :
+        dateFormatter(date, locale, dateTimezoneOffset) :
```
getDateFromatter ?

ocombe replied:
no, `dateFormatter` that we define just above

vicb replied:
I mean the name could be improved

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -149,255 +149,331 @@ enum LocaleData {
+/**
+ * Use this enum to find the index of each type of locale data from the extra ngLocale data array
+ */
+enum ExtraLocaleData {
```
-> better name

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -472,7 +555,31 @@ function toCamelCase(str: string): string {
+  const localeId = toCamelCase(data[LocaleData.LocaleId]);
+  LOCALE_DATA[localeId] = data;
+  if (extraData) {
+    LOCALE_DATA[localeId][LocaleData.DayPeriodsRules] = extraData[ExtraLocaleData.DayPeriodsRules];
```
why `DayPeriodsRules` while other are prefixed with `Extra` ?
why multiple slots ?

mary-poppins commented:
You can preview 252bbb4 at https://pr18284-252bbb4.ngbuilds.io/.

mary-poppins commented:
You can preview 5bca539 at https://pr18284-5bca539.ngbuilds.io/.

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -460,18 +458,25 @@ export function getLocaleDayPeriodRules(locale: string): (Time | [Time, Time])[]
 
 /**
- * Returns the first value that is defined in an array, going backwards
+ * Returns the first value that is defined in an array, going backwards.
```
first ... backwards = last ? -> `getLastDefinedValue` ?

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -460,18 +458,25 @@ export function getLocaleDayPeriodRules(locale: string): (Time | [Time, Time])[]
- * Returns the first value that is defined in an array, going backwards
+ * Returns the first value that is defined in an array, going backwards.
+ *
+ * To avoid repeating equivalent data (e.g. when "format" and "standalone" are the same) we only
```
`equivalent` -> `same` ?

vicb replied:
"the other ones are undefined if they are duplicates" -> "are only defined when different"

vicb commented on packages/common/test/pipes/date_pipe_spec.ts:
```diff
@@ -271,11 +271,12 @@ export function main() {
       it('should format correctly in IE11',
          () => expect(pipe.transform('2017-05-07T22:14:39', 'dd-MM-yyyy HH:mm'))
-                   .toEqual('07-05-2017 22:14'));
+                   .toMatch(/07-05-2017 \d{2}:\d{2}/));
```
why change from `22:14` to `\d{2}:\d{2}` ? Is there a change in TZ handling ?

ocombe replied:
we're testing the format, not the timezone (and it failed on travis because of the different time zone)

vicb replied:
Do you understand why ?
Why do we have a "in IE11" remaining in this spec ?

ocombe replied:
I removed the ie11 test

vicb commented on packages/common/test/pipes/date_pipe_spec.ts:
```diff
@@ -271,11 +271,12 @@ export function main() {
       it('should show the correct time in Safari with shortTime and custom format', () => {
-        expect(pipe.transform('2017-06-13T10:14:39', 'shortTime')).toEqual('10:14 AM');
-        expect(pipe.transform('2017-06-13T10:14:39', 'h:mm a')).toEqual('10:14 AM');
+        expect(pipe.transform('2017-06-13T10:14:39+0000', 'shortTime', '+0000'))
```
why changed ?

ocombe replied:
it failed the tests on travis (worked fine on local)

vicb replied:
Do you understand why ?
Why do we have a "in Safari" remaining in this spec ?

ocombe replied:
The problem is explained in the deprecated date pipe test with the following comment:
```ts
// TODO: reactivate the disabled expectations once emulators are fixed in SauceLabs
// In some old versions of Chrome in Android emulators, time formatting returns dates in the
// timezone of the VM host,
// instead of the device timezone. Same symptoms as
// https://bugs.chromium.org/p/chromium/issues/detail?id=406382
// This happens locally and in SauceLabs, so some checks are disabled to avoid failures.
// Tracking issue: https://github.com/angular/angular/issues/11187
```

The original tests for IE & safari only checks the date, not the hours/minutes because of that bug:
```ts
it('should format invalid in Safari ISO date',
         () => expect(pipe.transform('2017-01-20T19:00:00+0000')).toEqual('Jan 20, 2017'));
```

What we are testing here is that we can actually parse an ISO string as is it formatted by safari/IE (which is not a "valid" ISO string as defined in the ISO spec)

I'm not sure why I changed the test to minutes/hours, I think that what I wanted to do originally is test whether fixing the new "zone" parameter fixes the problem of mismatched hours/minutes (which is the point of this parameter).
I'm going to reintroduce the 2 original safari/IE test, and add a new one that checks the timezone parameter.

ocombe replied:
Ok I reran the tests without the timezone and we have the bug in safari: https://travis-ci.org/angular/angular/jobs/264651270#L869-L912
This is the bug reported here: https://github.com/angular/angular/issues/17478
I'm not sure how we should fix this

mary-poppins commented:
You can preview 8513413 at https://pr18284-8513413.ngbuilds.io/.

mary-poppins commented:
You can preview 1da05f9 at https://pr18284-1da05f9.ngbuilds.io/.

mary-poppins commented:
You can preview 764e74a at https://pr18284-764e74a.ngbuilds.io/.

mary-poppins commented:
You can preview a6dc69f at https://pr18284-a6dc69f.ngbuilds.io/.

mary-poppins commented:
You can preview cc8d7e4 at https://pr18284-cc8d7e4.ngbuilds.io/.

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -178,10 +178,10 @@ export function getLocaleId(locale: string): string {
     locale: string, formStyle: FormStyle, width: TranslationWidth): [string, string] {
   const data = findLocaleData(locale);
-  const amPm = firstDefined<[string, string][]>(
+  const amPm = getLastDefinedValue<[string, string][]>(
```
remove the type on call sites `getLastDefinedValue(...)`

vicb commented on packages/common/test/pipes/date_pipe_spec.ts:
```diff
@@ -269,14 +269,19 @@ export function main() {
+      // test for the following bugs:
+      // https://github.com/angular/angular/issues/16624
+      // https://github.com/angular/angular/issues/17478
+      xit('should show the correct time in Safari with shortTime and custom format', () => {
```
xit ?

mary-poppins commented:
You can preview f10f254 at https://pr18284-f10f254.ngbuilds.io/.

mary-poppins commented:
You can preview 940cedf at https://pr18284-940cedf.ngbuilds.io/.

mary-poppins commented:
You can preview 57ebe23 at https://pr18284-57ebe23.ngbuilds.io/.

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -78,8 +78,8 @@ export function formatNumber(
         maxFraction = parseIntAutoRadix(parts[5]);
-      } else if (parts[3] != null) {
-        maxFraction = parseIntAutoRadix(parts[3]);
+      } else if (parts[3] != null && minFraction > maxFraction) {
```
PLS remove magic 1, 3, 5, replace with constants with comments on them

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -177,11 +177,11 @@ export function getLocaleId(locale: string): string {
     locale: string, formStyle: FormStyle, width: TranslationWidth): [string, string] {
-  const data = findLocaleData(locale);
-  const amPm = getLastDefinedValue<[string, string][]>(
+  const data = <[string, string][][]>findLocaleData(locale);
```
`const data = <[string, string][][]>findLocaleData(locale);`
vs
`const data = <string[][][]>findLocaleData(locale);`

at least one place is wrong.

Types are not here to (only) please the compiler !

ocombe replied:
they are both correct I think:
- the first one contains [string, string]
- the second one contains an unknown number of strings because of period rules (we can have 0+)

vicb commented on tools/gulp-tasks/cldr/extract.js:
```diff
@@ -436,6 +436,24 @@ function stringify(obj) {
+ *
+ * For further size improvements, "undefined" values will be replaced by empty values in the arrays
+ * as the last step of the file generation (in generateLocale and generateLocaleExtra).
+ * e.g.: [x, y, undefined, z, undefined, undefined] will be [x, y, , z, , ]
```
pls keep undefined here

ocombe replied:
this almost double the size of locale files and is not necessary, why would we keep it?

mary-poppins commented:
You can preview 9d537f6 at https://pr18284-9d537f6.ngbuilds.io/.

juleskremer reviewed (approved):
OC: Approved change to docs; however, additional changes will be needed post-RC0.  I will create a new issue when ready.

mary-poppins commented:
You can preview 2f8665a at https://pr18284-2f8665a.ngbuilds.io/.

mary-poppins commented:
You can preview f2ae5d0 at https://pr18284-f2ae5d0.ngbuilds.io/.

vicb commented on packages/common/src/pipes/date_pipe.ts:
```diff
@@ -7,66 +7,118 @@
+ *  |                    | EEEEE       | Narrow                                                        | T                                                          |
+ *  |                    | EEEEEE      | Short                                                         | Tu                                                         |
+ *  | Period             | a, aa & aaa | Abbreviated                                                   | am/pm or AM/PM                                             |
+ *  |                    | aaaa        | Wide                                                          | ante meridiem/post meridiem                                |
```
This does not match the code:
- remove the code,
- remove any provision there could be in the code for that

mary-poppins commented:
You can preview 146f211 at https://pr18284-146f211.ngbuilds.io/.

mary-poppins commented:
You can preview fe7d604 at https://pr18284-fe7d604.ngbuilds.io/.

mary-poppins commented:
You can preview 277aa56 at https://pr18284-277aa56.ngbuilds.io/.

mary-poppins commented:
You can preview c59f749 at https://pr18284-c59f749.ngbuilds.io/.

cexbrayat commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -0,0 +1,594 @@
+ * | exponential            | 1.2`E`3           | used in computers for 1.2×10³.              |
+ * | superscriptingExponent | 1.2`×`103         | human-readable format of exponential        |
+ * | infinity               | `∞`               | used in +∞ and -∞.                          |
+ * | nan                    | `NaN`             | "not a number".                             |
```
@ocombe Not sure if relevant: `timeSeparator` and `currencyGroup` are missing in the docs here.

ocombe replied:
good call, adding those, and the missing `currencyDecimal` also

cexbrayat commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -119,28 +79,40 @@ export class DecimalPipe implements PipeTransform {
- * - `symbolDisplay` is a boolean indicating whether to use the currency symbol or code.
- *   - `true`: use symbol (e.g. `$`).
- *   - `false`(default): use code (e.g. `USD`).
+ * - `display` indicates whether to show the currency symbol or the code.
```
@ocombe This is still called `symbolDisplay` in the `howToUse` section.

vicb commented:
LGTM when Travis is green.

@ocombe please rebase (the mem fix for CI has been merged), create a new issue for the optional test which is failing, add a link to the PR in the issue.
Thanks.

mary-poppins commented:
You can preview eac0b99 at https://pr18284-eac0b99.ngbuilds.io/.

mary-poppins commented:
You can preview 97c6974 at https://pr18284-97c6974.ngbuilds.io/.

mary-poppins commented:
You can preview 154def8 at https://pr18284-154def8.ngbuilds.io/.

mary-poppins commented:
You can preview 591dea5 at https://pr18284-591dea5.ngbuilds.io/.

mary-poppins commented:
You can preview 3ae3d61 at https://pr18284-3ae3d61.ngbuilds.io/.

mary-poppins commented:
You can preview 9613115 at https://pr18284-9613115.ngbuilds.io/.

vicb commented on packages/common/src/common.ts:
```diff
@@ -14,7 +14,7 @@
 export * from './location/index';
 export {NgLocaleLocalization, NgLocalization} from './i18n/localization';
 export {Plural, LOCALE_DATA} from './i18n/locale_data';
-export {findLocaleData, registerLocaleData, NumberFormatStyle, FormStyle, Time, TranslationWidth, FormatWidth, NumberSymbol, WeekDay, getLocaleDayPeriods, getLocaleDayNames, getLocaleMonthNames, getLocaleId, getLocaleEraNames, getLocaleWeekEndRange, getLocaleFirstDayOfWeek, getLocaleDateFormat, getLocaleDateTimeFormat, getLocaleExtraDayPeriodRules, getLocaleExtraDayPeriods, getLocalePluralCase, g ...
```
hum ?

ocombe committed:
feat(common): drop use of the Intl API to improve browser support

BREAKING CHANGE: Because of multiple bugs and browser inconsistencies, we have dropped the intl api in favor of data exported from the Unicode Common Locale Data Repository (CLDR).
Unfortunately we had to change the i18n pipes (date, number, currency, percent) and there are some breaking changes.

1. I18n pipes
* Breaking change:
  - By default Angular now only contains locale data for the language `en-US`, if you set the value of `LOCALE_ID` to another locale, you will have to import new locale data for this language because we don't use the intl API anymore.

* Features:
  - you don't need to use the intl polyfill for Angular anymore.
  - all i18n pipes now have an additional last parameter `locale` which allows you to use a specific locale instead of the one defined in the token `LOCALE_ID` (whose value is `en-US` by default).
  - the new locale data extracted from CLDR are now available to developers as well and can be used through an API (which should be especially useful for library authors).
  - you can still use the old pipes for now, but their names have been changed and they are no longer included in the `CommonModule`. To use them, you will have to import the `DeprecatedI18NPipesModule` after the `CommonModule` (the order is important):

  ```ts
  import { NgModule } from '@angular/core';
  import { CommonModule, DeprecatedI18NPipesModule } from '@angular/common';

  @NgModule({
    imports: [
      CommonModule,
      // import deprecated module after
      DeprecatedI18NPipesModule
    ]
  })
  export class AppModule { }
  ```

  Dont forget that you will still need to import the intl API polyfill if you want to use those deprecated pipes.


2. Date pipe
* Breaking changes:
  - the predefined formats (`short`, `shortTime`, `shortDate`, `medium`, ...) now use the patterns given by CLDR (like it was in AngularJS) instead of the ones from the intl API. You might notice some changes, e.g. `shortDate` will be `8/15/17` instead of `8/15/2017` for `en-US`.
  - the narrow version of eras is now `GGGGG` instead of `G`, the format `G` is now similar to `GG` and `GGG`.
  - the narrow version of months is now `MMMMM` instead of `L`, the format `L` is now the short standalone version of months.
  - the narrow version of the week day is now `EEEEE` instead of `E`, the format `E` is now similar to `EE` and `EEE`.
  - the timezone `z` will now fallback to `O` and output `GMT+1` instead of the complete zone name (e.g. `Pacific Standard Time`), this is because the quantity of data required to have all the zone names in all of the existing locales is too big.
  - the timezone `Z` will now output the ISO8601 basic format, e.g. `+0100`, you should now use `ZZZZ` to get `GMT+01:00`.

  | Field type | Format        | Example value         | v4 | v5            |
  |------------|---------------|-----------------------|----|---------------|
  | Eras       | Narrow        | A for AD              | G  | GGGGG         |
  | Months     | Narrow        | S for September       | L  | MMMMM         |
  | Week day   | Narrow        | M for Monday          | E  | EEEEE         |
  | Timezone   | Long location | Pacific Standard Time | z  | Not available |
  | Timezone   | Long GMT      | GMT+01:00             | Z  | ZZZZ          |

* Features
  - new predefined formats `long`, `full`, `longTime`, `fullTime`.
  - the format `yyy` is now supported, e.g. the year `52` will be `052` and the year `2017` will be `2017`.
  - standalone months are now supported with the formats `L` to `LLLLL`.
  - week of the year is now supported with the formats `w` and `ww`, e.g. weeks `5` and `05`.
  - week of the month is now supported with the format `W`, e.g. week `3`.
  - fractional seconds are now supported with the format `S` to `SSS`.
  - day periods for AM/PM now supports additional formats `aa`, `aaa`, `aaaa` and `aaaaa`. The formats `a` to `aaa` are similar, while `aaaa` is the wide version if available (e.g. `ante meridiem` for `am`), or equivalent to `a` otherwise, and `aaaaa` is the narrow version (e.g. `a` for `am`).
  - extra day periods are now supported with the formats `b` to `bbbbb` (and `B` to `BBBBB` for the standalone equivalents), e.g. `morning`, `noon`, `afternoon`, ....
  - the short non-localized timezones are now available with the format `O` to `OOOO`. The formats `O` to `OOO` will output `GMT+1` while the format `OOOO` will be `GMT+01:00`.
  - the ISO8601 basic time zones are now available with the formats `Z` to `ZZZZZ`. The formats `Z` to `ZZZ` will output `+0100`, while the format `ZZZZ` will be `GMT+01:00` and `ZZZZZ` will be `+01:00`.

* Bug fixes
  - the date pipe will now work exactly the same across all browsers, which will fix a lot of bugs for safari and IE.
  - eras can now be used on their own without the date, e.g. the format `GG` will be `AD` instead of `8 15, 2017 AD`.


3. Currency pipe
* Breaking change:
  - the default value for `symbolDisplay` is now `symbol` instead of `code`. This means that by default you will see `$4.99` for `en-US` instead of `USD4.99` previously.
  
* Deprecation:
  - the second parameter of the currency pipe (`symbolDisplay`) is no longer a boolean, it now takes the values `code`, `symbol` or `symbol-narrow`. A boolean value is still valid for now, but it is deprecated and it will print a warning message in the console.

* Features:
  - you can now choose between `code`, `symbol` or `symbol-narrow` which gives you access to more options for some currencies (e.g. the canadian dollar with the code `CAD` has the symbol `CA$` and the symbol-narrow `$`).


4. Percent pipe
* Breaking change
  - if you don't specify the number of digits to round to, the local format will be used (and it usually rounds numbers to 0 digits, instead of not rounding previously), e.g. `{{ 3.141592 | percent }}` will output `314%` for the locale `en-US` instead of `314.1592%` previously.


Fixes #10809, #9524, #7008, #9324, #7590, #6724, #3429, #17576, #17478, #17319, #17200, #16838, #16624, #16625, #16591, #14131, #12632, #11376, #11187
